pandas/_libs/tslibs/parsing.pyx

```diff
@@ -0,0 +1,688 @@
+                      yearfirst=yearfirst, **kwargs)
+        return dt
+    try:
+        dt, _, _ = _parse_dateabbr_string(date_string, _DEFAULT_DATETIME, freq)
+        return dt
+    except DateParseError:
+        raise
+    except ValueError:
+        pass
+
+    try:
+        dt = du_parse(date_string, default=_DEFAULT_DATETIME,
+                      dayfirst=dayfirst, yearfirst=yearfirst, **kwargs)
+    except TypeError:
+        # following may be raised from dateutil
+        # TypeError: 'NoneType' object is not iterable
+        raise ValueError('Given date string not likely a datetime.')
+
+    return dt
+
+
+# Moved from pandas.core.tools.datetimes
+def parse_time_string(arg, freq=None, dayfirst=None, yearfirst=None):
+    """
+    Try hard to parse datetime string, leveraging dateutil plus some extra
+    goodies like quarter recognition.
+
+    Parameters
+    ----------
+    arg : compat.string_types
+    freq : str or DateOffset, default None
+        Helps with interpreting time string if supplied
+    dayfirst : bool, default None
+        If None uses default from print_config
+    yearfirst : bool, default None
+        If None uses default from print_config
+
+    Returns
+    -------
+    datetime, datetime/dateutil.parser._result, str
+    """
+    
+    if not isinstance(arg, string_types):
+        return arg
+
+    if getattr(freq, "_typ", None) == "dateoffset":
+        freq = freq.rule_code
+
+    if dayfirst is None:
+        from pandas.core.config import get_option
+        dayfirst = get_option("display.date_dayfirst")
+    if yearfirst is None:
+        from pandas.core.config import get_option
+        yearfirst = get_option("display.date_yearfirst")
+
+    return parse_datetime_string_with_reso(arg, freq=freq,
+                                                 dayfirst=dayfirst,
+                                                 yearfirst=yearfirst)
+
+
+@cython.locals(date_string=object, freq=object)
+def parse_datetime_string_with_reso(date_string, freq=None, dayfirst=False,
+                                    yearfirst=False, **kwargs):
+    """parse datetime string, only returns datetime
+
+    Returns
+    -------
+    datetime
+    """
+
+    cdef:
+        object parsed, reso
+
+    if not _does_string_look_like_datetime(date_string):
+        raise ValueError('Given date string not likely a datetime.')
+
+    try:
+        return _parse_dateabbr_string(date_string, _DEFAULT_DATETIME, freq)
+    except DateParseError:
+        raise
+    except ValueError:
+        pass
+
+    try:
+        parsed, reso = dateutil_parse(date_string, _DEFAULT_DATETIME,
+                                      dayfirst=dayfirst, yearfirst=yearfirst)
+    except Exception as e:
+        # TODO: allow raise of errors within instead
+        raise DateParseError(e)
+    if parsed is None:
+        raise DateParseError("Could not parse %s" % date_string)
+    return parsed, parsed, reso
+
+
+@cython.returns(cython.bint)
```

Comment:
these are not typical, again you don't need to type everything. if its a perf issue, pls demonstrate it.

Reply:
The existing version is in `tslib`, has this typing information, though not using the decorators.  I'll revert from pure-python-mode to be identical to existing for now.